# Pandas Tutorial

Author: Kellen Sullivan

This is an introductory tutorial for the popular data handling python package pandas. If you are unfamiliar with NumPy, you should complete the NumPy tutorial before completing this tutorial.

## DataFrames

DataFrames are the essential underlying datastructure for storing data with pandas. They are tables, with labeled columns and indexed rows. Series are a similar datatype in pandas, but contain only one dimension. Because of this, Series do not have labeled columns, and instead only have indexed rows. 

To get started using pandas, you must first import the pandas package. To do this we will use the following line of code:

# THIS DOESN'T MAKE SENSE BECUASE WHEN I DEFINE A SERIES BELOW, IT HAS A LABELED COLUMN. I NEED TO WORK ON THIS INTRODUCTORY SECTION MORE!!!

In [62]:
import pandas as pd

Note that `as pd` imports pandas with the alias `pd`. This saves us time from typing out `pandas` everytime we want to refer to the package and instead we can type `pd`. 

### 1. Create a Dataframe

In [ ]:
df = pd.DataFrame(
    {
        'Color': ["red", "blue", "green", "purple", "white", "orange"],
        'Price' : [5, 8, 3, 4, 9, 5]
    }
)

# display the type of a Pandas DataFrame
type(df)

pandas.core.frame.DataFrame

### 2. Create a Series 

In [ ]:
s = pd.Series(
    {
        'Color': ["red", "blue", "green", "purple", "white"],
    }
)

# display the type of a Pandas Series
type(s)

Color    [red, blue, green, purple, white]
dtype: object


pandas.core.series.Series

## Load and Explore DataFrames

As displayed above, DataFrames can be created from python dictionaries. However, when working with real-world data you will often want to work with data that is stored in another location. Fortunately, pandas provides many built in functions to easily load data from a variety of file types. 

For this tutorial, we will explore a popular real-world dataset within the machine learning community about the passengers on the Titanic. The data is stored in a CSV  (Comma-Separated Values) file, which pandas is well suited to handle.

Use `read_csv()` to read in data from a csv file and store it into a Pandas DataFrame. Include the filepath to the csv file to read. 

We also include `index_col="PassengerId"` to set the index of the DataFrame to be the PassengerId column.

In [63]:
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv", index_col="PassengerId")

Use `head()` to display the first 5 rows of a DataFrame. You can also provide a value n to display the first n rows. For example `head(20)` displays the first 20 rows of a DataFrame.

In [3]:
# default displays the first 5 rows of a DataFrame
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

The tail function works similarly to the head function, but displays the last 5 rows of a DataFrame

In [81]:
df.tail()

Survived  Pclass                                      Name  \
PassengerId                                                               
887                 0       2                     Montvila, Rev. Juozas   
888                 1       1              Graham, Miss. Margaret Edith   
889                 0       3  Johnston, Miss. Catherine Helen "Carrie"   
890                 1       1                     Behr, Mr. Karl Howell   
891                 0       3                       Dooley, Mr. Patrick   

                Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
PassengerId                                                                
887            male  27.0      0      0      211536  13.00   NaN        S  
888          female  19.0      0      0      112053  30.00   B42        S  
889          female   NaN      1      2  W./C. 6607  23.45   NaN        S  
890            male  26.0      0      0      111369  30.00  C148        C  
891            male  32.0      0      0      370376   7.75   NaN        Q

DataFrames have a few attributes that allow for data scientist to quickly view general information about a data set. To invoke an attribute use the syntax `dfname.attribute`

`.dtypes` displays all columns in a DataFrame and their corresponding data type

In [ ]:
print(df.dtypes)

Survived      int64
Pclass        int64
Name         object
Sex          object
Age         float64
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Cabin        object
Embarked     object
dtype: object


`.shape` returns a tuple with the number of rows followed by the number of columns

In [83]:
print(df.shape)

(891, 11)


`.size` returns the total number of elements in a DataFrame. In other words, it returns the product of the number of rows and columns in a DataFrame. 

In [84]:
print(df.size)

9801


## Selecting Data

Pandas provides multiple ways to select subsets of data within a DataFrame. In this tutorial we will go over the following methods:

- bracket and dot notation
- loc and iloc
- query

### bracket and dot notation

The simplest way to select a subset of data is similar to selecting an element in a standard python list or dictionary.

- To select a column, use `[]` with the column name inside.
- You can also use a `.` followed by the column name. 

Each syntax will produce the same result

In [11]:
print(df["Age"])

PassengerId
1      22.0
2      38.0
3      26.0
4      35.0
5      35.0
       ... 
887    27.0
888    19.0
889     NaN
890    26.0
891    32.0
Name: Age, Length: 891, dtype: float64


In [12]:
print(df.Age)

PassengerId
1      22.0
2      38.0
3      26.0
4      35.0
5      35.0
       ... 
887    27.0
888    19.0
889     NaN
890    26.0
891    32.0
Name: Age, Length: 891, dtype: float64


When using bracket notation, you can pass in a list to select multiple columns. This is one advantage of using bracket notation. However, both are valid and the best one to use depends on the context.

In [108]:
columns_to_keep = ["Sex", "Age", "Survived"]

print(df[columns_to_keep])

                Sex   Age  Survived
PassengerId                        
1              male  22.0         0
2            female  38.0         1
3            female  26.0         1
4            female  35.0         1
5              male  35.0         0
...             ...   ...       ...
887            male  27.0         0
888          female  19.0         1
889          female   NaN         0
890            male  26.0         1
891            male  32.0         0

[891 rows x 3 columns]


### loc and iloc

Pandas provides `loc` and `iloc` as indexing options for more advanced operations. `loc` and `iloc` are invoked similarly, with the row selector first followed by the column selector

`loc`  syntax: `df.loc[row_labels, column_labels]`  
`iloc` syntax: `df.iloc[row_positions, column_positions]`

You can also provide only one selector to `loc` or `iloc`. Pandas will assume you are only selecting based on row and all columns will be included. For example:

`df.loc[row_labels]` is also valid syntax

The main distinction between `loc` and `iloc` is that `loc` uses label-based selection and `iloc` uses position-based selection.

To select an element from a DataFrame using `loc`, provide the row label and column labels to select from. 

In [ ]:
# Selects the row with PassengerId of 3
df.loc[3]

# Selects the row with PassengerId of 3, and the columns Name and Sex
df.loc[3, ["Name", "Sex"]]

# Selects the rows with PassengerId's 3, 4, and 5 and the Age column
df.loc[[3,4,5], "Age"]

PassengerId
3    26.0
4    35.0
5    35.0
Name: Age, dtype: float64

To select the same elements using `iloc` we use the index instead of labels. 

Note that iloc uses 0-based indexing, so to select the third row in the DataFrame, which is the row with PassengerId of 3, we use 2. 

In [ ]:
# Selects the third row in the DataFrame
df.iloc[2]

# Selects the third row, and the third and fourth columns in the DataFrame
df.iloc[2, [2,3]]

# Selects the third, fourth, and fifth rows and the fifth column in the DataFrame
df.iloc[[2,3,4], 4]

PassengerId
3    26.0
4    35.0
5    35.0
Name: Age, dtype: float64

You can use the slicing operator `:` to select many rows or columns without listing all of them individually. The following statement selects rows with PassengerId's from 3 to 10, and all columns up until Age.

In [32]:
df.loc[3:10, :"Age"]

Survived  Pclass  \
PassengerId                     
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   

                                                          Name     Sex   Age  
PassengerId                                                                   
3                                       Heikkinen, Miss. Laina  female  26.0  
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  
5                                     Allen, Mr. William Henry    male  35.0  
6                                             Moran, Mr. James    male   NaN  
7                                      McCarthy, Mr. Timothy J    male  54.0  
8                               Palsson, Master. Gosta Leonard    male   2.0  
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0  
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0

Using iloc, a similar subset of the DataFrame can be selected. However, due to iloc's 0-based indexing, the row with PassengerId of 3 is not included!

In [33]:
df.iloc[3:10, :5]

Survived  Pclass  \
PassengerId                     
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   

                                                          Name     Sex   Age  
PassengerId                                                                   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0  
5                                     Allen, Mr. William Henry    male  35.0  
6                                             Moran, Mr. James    male   NaN  
7                                      McCarthy, Mr. Timothy J    male  54.0  
8                               Palsson, Master. Gosta Leonard    male   2.0  
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0  
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0

One advantage of using `loc` is that it can handle conditional selection. To select based on a conditional statement use the syntax:

`df.loc[conditional statement]`

In [ ]:
# Select all rows with Age < 18
df.loc[df["Age"] < 18]

Survived  Pclass                                     Name  \
PassengerId                                                              
8                   0       3           Palsson, Master. Gosta Leonard   
10                  1       2      Nasser, Mrs. Nicholas (Adele Achem)   
11                  1       3          Sandstrom, Miss. Marguerite Rut   
15                  0       3     Vestrom, Miss. Hulda Amanda Adolfina   
17                  0       3                     Rice, Master. Eugene   
...               ...     ...                                      ...   
851                 0       3  Andersson, Master. Sigvard Harald Elias   
853                 0       3                  Boulos, Miss. Nourelain   
854                 1       1                Lines, Miss. Mary Conover   
870                 1       3          Johnson, Master. Harold Theodor   
876                 1       3         Najib, Miss. Adele Kiamie "Jane"   

                Sex   Age  SibSp  Parch    Ticket     Fare Cabin Embarked  
PassengerId                                                                
8              male   2.0      3      1    349909  21.0750   NaN        S  
10           female  14.0      1      0    237736  30.0708   NaN        C  
11           female   4.0      1      1   PP 9549  16.7000    G6        S  
15           female  14.0      0      0    350406   7.8542   NaN        S  
17             male   2.0      4      1    382652  29.1250   NaN        Q  
...             ...   ...    ...    ...       ...      ...   ...      ...  
851            male   4.0      4      2    347082  31.2750   NaN        S  
853          female   9.0      1      1      2678  15.2458   NaN        C  
854          female  16.0      0      1  PC 17592  39.4000   D28        S  
870            male   4.0      1      1    347742  11.1333   NaN        S  
876          female  15.0      0      0      2667   7.2250   NaN        C  

[113 rows x 11 columns]

Often `loc` and `iloc` are both viable solutions, so deciding when to use which one can be tricky. In general, if the DataFrame has clear row and column lables, or you are using conditional selection, then `loc` is the better option. If instead label names are not relevant, or you are iterating over a DataFrame, then `iloc` may provide a better solution.

### query

`query()` is a useful function that provides another way to select elements in a DataFrame based on a condition. For example, you can select all elements in the DataFrame where the passenger is a male and survived using the following statement.

In [51]:
df.query("Sex == 'male' and Survived == 1")

Survived  Pclass                             Name   Sex   Age  \
PassengerId                                                                  
18                  1       2     Williams, Mr. Charles Eugene  male   NaN   
22                  1       2            Beesley, Mr. Lawrence  male  34.0   
24                  1       1     Sloper, Mr. William Thompson  male  28.0   
37                  1       3                 Mamee, Mr. Hanna  male   NaN   
56                  1       1                Woolner, Mr. Hugh  male   NaN   
...               ...     ...                              ...   ...   ...   
839                 1       3                  Chip, Mr. Chang  male  32.0   
840                 1       1             Marechal, Mr. Pierre  male   NaN   
858                 1       1           Daly, Mr. Peter Denis   male  51.0   
870                 1       3  Johnson, Master. Harold Theodor  male   4.0   
890                 1       1            Behr, Mr. Karl Howell  male  26.0   

             SibSp  Parch  Ticket     Fare Cabin Embarked  
PassengerId                                                
18               0      0  244373  13.0000   NaN        S  
22               0      0  248698  13.0000   D56        S  
24               0      0  113788  35.5000    A6        S  
37               0      0    2677   7.2292   NaN        C  
56               0      0   19947  35.5000   C52        S  
...            ...    ...     ...      ...   ...      ...  
839              0      0    1601  56.4958   NaN        S  
840              0      0   11774  29.7000   C47        C  
858              0      0  113055  26.5500   E17        S  
870              1      1  347742  11.1333   NaN        S  
890              0      0  111369  30.0000  C148        C  

[109 rows x 11 columns]

To refer to variables within a query statement, use `@` followed by the variable name

In [52]:
cutoff_age = 30

df.query("Age < @cutoff_age")

Survived  Pclass  \
PassengerId                     
1                   0       3   
3                   1       3   
8                   0       3   
9                   1       3   
10                  1       2   
...               ...     ...   
884                 0       2   
885                 0       3   
887                 0       2   
888                 1       1   
890                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
...                                                        ...     ...   ...   
884                              Banfield, Mr. Frederick James    male  28.0   
885                                     Sutehall, Mr. Henry Jr    male  25.0   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
890                                      Behr, Mr. Karl Howell    male  26.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
8                3      1            349909  21.0750   NaN        S  
9                0      2            347742  11.1333   NaN        S  
10               1      0            237736  30.0708   NaN        C  
...            ...    ...               ...      ...   ...      ...  
884              0      0  C.A./SOTON 34068  10.5000   NaN        S  
885              0      0   SOTON/OQ 392076   7.0500   NaN        S  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
890              0      0            111369  30.0000  C148        C  

[384 rows x 11 columns]

Both `query` and `loc` can select subsets of a DataFrame based on a condition. The advantage to using `query` is that it is often quicker to write and easier to read. However, for more advanced conditional statements with Python logic/methods using `loc` is necessary.

### Exercise: Perform Exploratory Data Analysis

**Time to put your knowledge to the test!**

An important first step in any machine learning project is to explore the data for any intial insights. Pandas provides various functions machine learning engineers in this task.

For this exercise you will explore a very popular dataset about passengers on the Titanic. In order to complete this exercise you must complete the following:
1. Display the first 10 rows of the DataFrame
2. Determine the amount of rows and columns in the DataFrame
3. Display all columns and their datatypes
4. Create a subset of the original DataFrame that contains all passengers that are female OR under 18 years old

In [137]:
# 1. Display the first 10 rows of the DataFrame


In [ ]:
# 2. Determine the amount of rows and columns in the DataFrame


# 3. Display all columns and their datatypes



In [ ]:
# 4. Create a DataFrame containing only passengers that are either female OR 18 years or younger
sub_df = 

sub_df.head(10)

#### **Solution**

(Try to solve it yourself first before looking!)

Click the cells below to reveal a possible solution approach.

In [87]:
# 1. Display the first 10 rows of the DataFrame
df.head(10)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
6                0      0            330877   8.4583   NaN        Q  
7                0      0             17463  51.8625   E46        S  
8                3      1            349909  21.0750   NaN        S  
9                0      2            347742  11.1333   NaN        S  
10               1      0            237736  30.0708   NaN        C

In [134]:
# 2. Determine the amount of rows and columns in the DataFrame
print(df.shape)

# 3. Display all columns and their datatypes
print(df.dtypes)

(891, 11)
Survived      int64
Pclass        int64
Name         object
Sex          object
Age         float64
SibSp         int64
Parch         int64
Ticket       object
Fare        float64
Cabin        object
Embarked     object
dtype: object


In [ ]:
# 4. Create a DataFrame containing only female passengers 18 or younger
sub_df = df.query("Sex == 'female' or Age < 18")
sub_df = df.loc[(df["Sex"] == "female") | (df["Age"] < 18)] # Alternative solution using loc

sub_df.head(10)

Survived  Pclass  \
PassengerId                     
2                   1       1   
3                   1       3   
4                   1       1   
8                   0       3   
9                   1       3   
10                  1       2   
11                  1       3   
12                  1       1   
15                  0       3   
16                  1       2   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
15                        Vestrom, Miss. Hulda Amanda Adolfina  female  14.0   
16                            Hewlett, Mrs. (Mary D Kingcome)   female  55.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
8                3      1            349909  21.0750   NaN        S  
9                0      2            347742  11.1333   NaN        S  
10               1      0            237736  30.0708   NaN        C  
11               1      1           PP 9549  16.7000    G6        S  
12               0      0            113783  26.5500  C103        S  
15               0      0            350406   7.8542   NaN        S  
16               0      0            248706  16.0000   NaN        S

## Aggregation 

Pandas provides many functions to quickly summarize or aggregate data within DataFrames. In this section we will discuss the following:
- Aggregate Operations
- groupby
- value counts

### Aggregate Operations
Pandas provides many aggregate operations to quickly summarize data into a single value. Aggregate operations iterate over a column and return a single value. The following syntax is used to use one:
`df['column_name'].aggregate_operation()`

If an aggregate operation is used on a DataFrame with more than one column, Pandas will apply the operation to each column in the DataFrame and return a Series containing the result of applying the operation to each column in the DataFrame.

Some of the most useful operations include:
- mean
- sum
- median
- count
- max
- min

In [60]:
# Returns the average age of a passenger
print(f"Average Age: {df['Age'].mean()}")

# Adds the Survived value for every row in the DataFrame
# Because a 1 indicates a passenger survived and 0 indicates they did not
# The sum of the Survived columns is also the number of survivors
print(f"Number of Survivors: {df['Survived'].sum()}")

# Returns the median age of a passenger
print(f"Median Age: {df['Age'].median()}")

# Returns the number of non-null Name's in the DataFrame
print(f"Number of non-null Names: {df['Name'].count()}")

# Returns the greatest value in the Fare column
print(f"Most expensive Fare: {df['Fare'].max()}")

# Returns the least value in the Fare column
print(f"Least expensive Fare: {df['Fare'].min()}")

Average Age: 29.69911764705882
Number of Survivors: 342
Median Age: 28.0
Number of non-null Names: 891
Most expensive Fare: 512.3292
Least expensive Fare: 0.0


### groupby

Pandas provides the groupby function to split a DataFrame into distinct groups based on a key (usually an existing column in the DataFrame). You can then apply a function to each group individually. The resulting groups are then combined back into one DataFrame. This process is known as the split -> apply -> combine pattern that is commonly used in data analysis.

For example, the following code splits the DataFrame based on the column 'Sex'. It then applys the `sum()` function to the male group and female group individually. The results are then displayed in one DataFrame,

In [74]:
df.groupby('Sex')['Survived'].sum()

Sex
female    233
male      109
Name: Survived, dtype: int64

### Value Counts

`value_counts()` is another useful function for displaying the frequency of unqiue values in a DataFrame. For example, to see how many of the passengers are male and how many are female we can use the following code.

In [65]:
df["Sex"].value_counts()

Sex
male      577
female    314
Name: count, dtype: int64

The value counts function is often used in combination with the groupby function to display more complex relationships. In this situation, `value_counts()` acts as the function in the apply step in the split -> apply -> combine patter, and is applied to each group individually. 

In [70]:
df.groupby("Sex")["Survived"].value_counts()

Sex     Survived
female  1           233
        0            81
male    0           468
        1           109
Name: count, dtype: int64

### Exercise: Utilize Aggregations

**Put your aggregation knowledge to the test!**

Now that you have already intially explored the dataset, it is time to use aggregations to uncover deeper insights and better understand the data. In order to complete this exercise you must complete the following:

1. Determine the age of the youngest and oldest passengers to survive
2. Determine the average fare price for each passenger class 
3. Determine the number of passengers in each passenger class 
4. Determine the number of survivors and non-survivors for each passenger class

hint: passenger class is denoted by the column Pclass



In [ ]:
# 1. Determine the age of the youngest and oldest passengers to survive
youngest_survivor = ____
print(f"Age of the youngest passenger to survive: {youngest_survivor}")

oldest_survivor = ____
print(f"Age of the oldest passenger to survive: {oldest_survivor}")

# 2. Determine the average fare price for each passenger class
avg_fare_price = ____
print(f"Average fare price for each passenger class: \n {avg_fare_price}")

# 3. Determine the number of passengers in each passenger class
passenger_count_by_pclass = ____
print(f"Number of passengers by each passenger class: \n {passenger_count_by_pclass}")

# 4. Determine the number of survivors and non-survivors for each passenger class
survivors_by_pclass = ____
print(f"Survivors by passenger class: \n {survivors_by_pclass}")

#### **Solution**

(Try to solve it yourself first before looking!)

Click the cells below to reveal a possible solution approach.

In [92]:
# 1. Determine the age of the youngest and oldest passengers to survive
youngest_survivor = df.query("Survived == 1")['Age'].min()  
youngest_survivor = df.loc[df["Survived"] == 1]["Age"].min() # Alternative solution using loc
print(f"Age of the youngest passenger to survive: {youngest_survivor}")

oldest_survivor = df.query("Survived == 1")['Age'].max()
oldest_survivor = df.loc[df['Survived'] == 1]["Age"].max() # Alternative solution using loc
print(f"Age of the oldest passenger to survive: {oldest_survivor}\n")

# 2. Determine the average fare price for each passenger class
avg_fare_price = df.groupby('Pclass')['Fare'].mean()
print(f"Average fare price for each passenger class: \n {avg_fare_price}\n")

# 3. Determine the number of passengers in each passenger class
passenger_count_by_pclass = df['Pclass'].value_counts()
print(f"Number of passengers by each passenger class: \n {passenger_count_by_pclass}\n")

# 4. Determine the number of survivors and non-survivors for each passenger class
survivors_by_pclass = df.groupby('Pclass')['Survived'].value_counts()
print(f"Survivors by passenger class: \n {survivors_by_pclass}")


Age of the youngest passenger to survive: 0.42
Age of the oldest passenger to survive: 80.0

Average fare price for each passenger class: 
 Pclass
1    84.154687
2    20.662183
3    13.675550
Name: Fare, dtype: float64

Number of passengers by each passenger class: 
 Pclass
3    491
1    216
2    184
Name: count, dtype: int64

Survivors by passenger class: 
 Pclass  Survived
1       1           136
        0            80
2       0            97
        1            87
3       0           372
        1           119
Name: count, dtype: int64


## Transforming Data

Data scientist will often recieve data in a state that is not suitable for machine learning models. Thankfully, Pandas provides many useful ways to transform and manipulate data into a form that is ideal for A.I. applications. In this section, we will review how to do the following: 

- cast columns to different types
- handle null values
- create new columns

### Type Casting
To change the data type of a column use `astype()`.
Common dtypes include 
- Int64
- Float64
- object
- string
- datetime64[ns]
- boolean  

Although there are many more!

In [ ]:
# Convert the name column from object to string
df['Name'].astype('string')

PassengerId
1                                Braund, Mr. Owen Harris
2      Cumings, Mrs. John Bradley (Florence Briggs Th...
3                                 Heikkinen, Miss. Laina
4           Futrelle, Mrs. Jacques Heath (Lily May Peel)
5                               Allen, Mr. William Henry
                             ...                        
887                                Montvila, Rev. Juozas
888                         Graham, Miss. Margaret Edith
889             Johnston, Miss. Catherine Helen "Carrie"
890                                Behr, Mr. Karl Howell
891                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: string

Note that this does not change the original DataFrame, but instead creates a copy of the DataFrame with the updated column type. In order to change the original DataFrame set the original DataFrame equal to the copy. (OR USE `copy=False` in the function)

In [ ]:
df['Name'] = df['Name'].astype('String')

### Address Null Values

If you are familiar with Python, you may know that null values are represented as `none`. In Pandas, `none` can be used to represent missing values, but there are also other representations. `NaN` (Not a number) indicates a missing numerical value. `NaT` (Not a time) is used to represent missing datime values. And finally `pd.NA` is used as a missing value indicator that applies to all data types. 

Most machine learning algorithms cannot directly work with null values, so it is important to have a strategy to handle them. 

In order to identify null values in a DataFrame, use `isna()`. This pandas function returns a Series or DataFrame with True values for every entry that is null

In [5]:
df.isna()

Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
PassengerId                                                                
1               False   False  False  False  False  False  False   False   
2               False   False  False  False  False  False  False   False   
3               False   False  False  False  False  False  False   False   
4               False   False  False  False  False  False  False   False   
5               False   False  False  False  False  False  False   False   
...               ...     ...    ...    ...    ...    ...    ...     ...   
887             False   False  False  False  False  False  False   False   
888             False   False  False  False  False  False  False   False   
889             False   False  False  False   True  False  False   False   
890             False   False  False  False  False  False  False   False   
891             False   False  False  False  False  False  False   False   

              Fare  Cabin  Embarked  
PassengerId                          
1            False   True     False  
2            False  False     False  
3            False   True     False  
4            False  False     False  
5            False   True     False  
...            ...    ...       ...  
887          False   True     False  
888          False  False     False  
889          False   True     False  
890          False  False     False  
891          False   True     False  

[891 rows x 11 columns]

The simplest strategy to handle null values is to remove or "drop" them. To drop all rows containing a null value use `dropna()`.

In [ ]:
dropped_df = df.dropna()

# 
dropped_df

Survived  Pclass  \
PassengerId                     
2                   1       1   
4                   1       1   
7                   0       1   
11                  1       3   
12                  1       1   
...               ...     ...   
872                 1       1   
873                 0       1   
880                 1       1   
888                 1       1   
890                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
...                                                        ...     ...   ...   
872           Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0   
873                                   Carlsson, Mr. Frans Olof    male  33.0   
880              Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
890                                      Behr, Mr. Karl Howell    male  26.0   

             SibSp  Parch    Ticket     Fare        Cabin Embarked  
PassengerId                                                         
2                1      0  PC 17599  71.2833          C85        C  
4                1      0    113803  53.1000         C123        S  
7                0      0     17463  51.8625          E46        S  
11               1      1   PP 9549  16.7000           G6        S  
12               0      0    113783  26.5500         C103        S  
...            ...    ...       ...      ...          ...      ...  
872              1      1     11751  52.5542          D35        S  
873              0      0       695   5.0000  B51 B53 B55        S  
880              0      1     11767  83.1583          C50        C  
888              0      0    112053  30.0000          B42        S  
890              0      0    111369  30.0000         C148        C  

[183 rows x 11 columns]

Dropping all rows with a null value can potentially exclude large amounts of valuable data. In this titanic example, `dropna()` reduced the number of rows in the DataFrame from 891 to 183!

Because of this, data scientist often prefer to "impute" or fill in missing values with some other value. One easy way to do this in pandas is to use `fillna()` and provide the value to replace null values with.

In [14]:
# replace all null values in the DataFrame with 0
df.head()
# df.fillna()

# may want to have something here to replace null values in a specific column

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

### Create New Columns

Pandas provides many ways to create new columns in a DataFrame.

To create a new column with a constant value use `df['new_column_name'] = val`. For example, the following code creates a new column 'Ship' in the DataFrame with the constant value 'Titanic'

In [11]:
df['Ship'] = 'Titanic'
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   

             Age_Centered     Ship  
PassengerId                         
1               -7.699118  Titanic  
2                8.300882  Titanic  
3               -3.699118  Titanic  
4                5.300882  Titanic  
5                5.300882  Titanic

It is also possible to create new columns based on existing columns in the DataFrame. The following code creates a new column where each row has a new value indicating the age distance from the mean age of everyone on board.

In [ ]:
df['Age_Centered'] = df['Age'] - df['Age'].mean()
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   

             Age_Centered  
PassengerId                
1               -7.699118  
2                8.300882  
3               -3.699118  
4                5.300882  
5                5.300882

Pandas also allows you to set new columns equal to a condition evaluated based on other columns. In this case, the value in the column 'Is_Female_Survivor' is true when the passenger survived AND they are a female. In this pandas expression we represent and using a single `&` symbol. Pandas uses the following operators for conditional statements.

- `&`  - and
- `|`  - or
- `==` - equal
- `!=` - notequal
- `>`  - greater than
- `<`  - less than
- `>=` - greater than or equal to
- `<=` - less than or equal to

In [38]:
df['Is_Female_Survivor'] = (df['Survived'] == 1) >= (df['Sex'] == 'female')
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   

             Is_Female_Survivor  
PassengerId                      
1                          True  
2                          True  
3                          True  
4                          True  
5                          True

The `apply()` function allows users to create new columns conditioned on the values of others. For example, the following code creates a new column 'Is_Child' that gets the value true when age is less than 18, and False otherwise. 

In [ ]:
def check_if_child(age):
    if age < 18:
        return True
    else:
        return False

# Not sure if I should use the lambda function, or define one explicity for clarity sake
df['Is_Child'] = df['Age'].apply(lambda x: True if x < 18 else False)
df.head(8)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   
6                0      0            330877   8.4583   NaN        Q   
7                0      0             17463  51.8625   E46        S   
8                3      1            349909  21.0750   NaN        S   

             Age_Centered     Ship  Is_Child  
PassengerId                                   
1               -7.699118  Titanic     False  
2                8.300882  Titanic     False  
3               -3.699118  Titanic     False  
4                5.300882  Titanic     False  
5                5.300882  Titanic     False  
6                     NaN  Titanic     False  
7               24.300882  Titanic     False  
8              -27.699118  Titanic      True

Map function here TODO!!

In [ ]:
df['Is_Child'] = df['Age'].apply(lambda x: True if x < 18 else False)
# maybe map function here as well!

### Exercise: Transform the DataFrame

**Time to put your knowledge to the test!**

Now that you have already explored the titanic dataset, it is time to TBD

For this exercise you will explore a very popular dataset about passengers on the Titanic. In order to complete this exercise you must complete the following:
1. Cast the Name column type to 'string'
2. Replace all null values in the 'Age' column with the mean age
3. Create a new column called 'Fare_centered' that is the difference of the Fare price for a passenger and the mean Fare price.
4. Create a new column called 'Adult_male' that is true when a passenger is a male and 18 years or older, and false otherwise


In [7]:
# 1. cast column type here


# 2. Replace all null values in the 'Age' column with the mean age


# 3. Create a new column called 'Fare_centered' that is the difference of the Fare price for a passenger and the mean Fare price.


# 4. Create a new column called 'Adult_male' that is true when a passenger is a male and 18 years or older, and false otherwise
    


#### **Solution**

(Try to solve it yourself first before looking!)

Click the cell below to reveal a possible solution approach.

In [11]:
# 1. cast column type here
df['Name'] = df['Name'].astype('string')

# 2. Replace all null values in the 'Age' column with the mean age
mean_age = df['Age'].mean()
df['Age'].fillna(mean_age)

# 3. Create a new column called 'Fare_centered' that is the difference of the Fare price for a passenger and the mean Fare price.
df['Fare_centered'] = df['Fare'] - df['Fare'].mean()

# 4. Create a new column called 'Adult_male' that is true when a passenger is a male and 18 years or older, and false otherwise
df['Is_Adult_Male'] = (df['Age'] >= 18) & (df['Sex'] == 'male')
df.head(10)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
6                   0       3   
7                   0       1   
8                   0       3   
9                   1       3   
10                  1       2   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
6                                             Moran, Mr. James    male   NaN   
7                                      McCarthy, Mr. Timothy J    male  54.0   
8                               Palsson, Master. Gosta Leonard    male   2.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  \
PassengerId                                                           
1                1      0         A/5 21171   7.2500   NaN        S   
2                1      0          PC 17599  71.2833   C85        C   
3                0      0  STON/O2. 3101282   7.9250   NaN        S   
4                1      0            113803  53.1000  C123        S   
5                0      0            373450   8.0500   NaN        S   
6                0      0            330877   8.4583   NaN        Q   
7                0      0             17463  51.8625   E46        S   
8                3      1            349909  21.0750   NaN        S   
9                0      2            347742  11.1333   NaN        S   
10               1      0            237736  30.0708   NaN        C   

             Fare_centered  Is_Adult_Male  
PassengerId                                
1               -24.954208           True  
2                39.079092          False  
3               -24.279208          False  
4                20.895792          False  
5               -24.154208           True  
6               -23.745908          False  
7                19.658292           True  
8               -11.129208          False  
9               -21.070908          False  
10               -2.133408          False

## Next Steps

Congratulations for completing the Pandas Introductory Tutorial! 🎓🎉

Here are some resources to futher your Pandas education:
- Official Pandas Documentation: https://pandas.pydata.org/docs/user_guide/10min.html
- Kaggle Pandas Tutorial: https://www.kaggle.com/learn/pandas
- Youtube Channel with various tutorials on Pandas and other Data Science topics: https://www.youtube.com/@robmulla

Polars is a modern DataFrame library that improves memory and efficieny from Pandas. Find their official documentation here:
https://docs.pola.rs/
